### PR TITLE
Vite: Re-export essentials sub-addon preview annotations for pnpm

### PR DESCRIPTION
**Pull request by IanVS**
Issue: #19176

Closes https://github.com/storybookjs/storybook/pull/19530

## What I did

In order to support pnpm from the vite builder, I added re-exports of the sub-addons which have preview annotations, and adjusted the preset handling to leave bare imports alone instead of turning them into absolute paths.  This allows pnpm to correctly find the files and avoids users from having to "shamefully hoist" or add the sub-addons to their package.json. 

I added `typesVersions` to the addons, so that we can `export * from "@storybook/addon-actions/preview"`, for instance.  The .d.ts file isn't at the root of the package, it's in `/dist`, so using typesVersions lets us point to the right spot.  This is one of the approaches shown in https://github.com/andrewbranch/example-subpath-exports-ts-compat, and has the best overall compatibility.

One problem I hit is that tsup generates empty .d.ts files for `manager`, so I filed https://github.com/egoist/tsup/issues/762 and needed to ignore errors when re-exporting from them.

## How to test

We don't have any good tests for pnpm, unfortunately, so this is a bit manual.

- Build this branch, especially `lib/builder-vite`, `lib/core-common`, and `addons/essentials`
- Somewhere else on your filesystem, create a pnpm vite project (`pnpm create vite@latest --template=react`)
- Add storybook (`pnpx sb@next init --builder=vite`)
- Copy over the `dist` of `lib/builder-vite` into `node_modules/.pnpm/@storybook+builder-vite@7.0.0-alpha.XX_<hash>/node_modules/@storybook/builder-vite`
- Copy over the `dist` of `lib/core-common` into `node_modules/.pnpm/@storybook+core-common@7.0.0-alpha.XX_<hash>/node_modules/@storybook/core-common`
- Copy the `dist` and `package.json` of `addons/essentials` into `node_modules/@storybook/addon-essentials/`
- Start up storybook, (`pnpm run storybook`), and it should start successfully.

### CODE REVIEW

**Comment by IanVS** [comment]
I'm currently getting an error of `Could not find a declaration file for module '@storybook/addon-actions/preview'`, though there's a clear types file declared in the export map for `@storybook/addon-actions`.  Not sure what's going on here.

**Comment by IanVS** [comment]
Blocked by https://github.com/storybookjs/storybook/pull/19735.

**Comment by tmeasday** [comment]
@IanVS If I am understanding the problem correctly here, then it's only the `previewAnnotations` where things are tricky. So:

```ts
export addons = [require.resolve('@storybook/addon-backgrounds'), ...]
```

Would work fine for everything except the `previewAnnotation` exported by the backgrounds addon's preview, which would end up being an absolute path, and thus not possible to be imported in Vite.

Another possibility, which I'm not sure if you've considered is adding an new exported field `addons` to `preview.ts`, so that we could add the following to `essentials/src/preview.ts:

```ts
import * as backgroundsAddon from '@storybook/addon-backgrounds/preview';

export addons = [backgroundsAddon, ... ];
```

Then the rule could be that a "addon that uses other addons" needs to do the above. And the preset handling could just stop looking at the `previewAnnotations` preset export after going one level deep. In theory we might need to do the same for the manager exports also.

**Comment by IanVS** [comment]
>Another possibility, which I'm not sure if you've considered is adding an new exported field addons to preview.ts, so that we could add the following to `essentials/src/preview.ts:

IIUC, I did explore something like that, in https://github.com/storybookjs/storybook/pull/19530, which did the actual imports and added them to the addons list.  But, that import would happen in node, and it needs to be transferred to the vite-builder virtual modules that run in the browser, which means serializing, which breaks since addons can export functions.

As for only changing the preview presets, I figured it would be simpler to avoid having some paths be absolute (manager) and some be bare imports (preview).  I think it's simpler to keep the same logic for manager and builder if possible.  Or do you see a downside to that?

File: code/addons/highlight/package.json
Comment (IanVS) [inline code/addons/highlight/package.json:67]: This change wasn't entirely necessary here, but seemed like a good idea to keep the addons consistent with each other.

File: code/addons/interactions/package.json
Comment (IanVS) [inline code/addons/interactions/package.json:102]: This could also have remained as before, but it was inconsistent with the rest of the addons, and it's nice for them all to follow the same pattern.  I did the same for `addon-outline` as well.

File: code/lib/builder-webpack5/package.json
Comment (IanVS) [inline code/lib/builder-webpack5/package.json:36]: These export maps were incorrect, pointing to the wrong folder, and using extensions, when we typically do not.

File: code/lib/docs-tools/src/hasDocsOrControls.ts
Comment (IanVS) [inline code/lib/docs-tools/src/hasDocsOrControls.ts:4]: This one had me scratching my head for half a day.  😆   I realized the Vite builder doesn't use this check, but probably should.

File: code/lib/core-common/src/presets.ts
Comment (IanVS) [inline code/lib/core-common/src/presets.ts:111]: We no longer need to turn these into absolute paths in the server, since we are no longer introducing addons that aren't in the user's package.json.  Additionally, this leaves it up to the consumer how to import/require the files, which means they'll get the right file (esm or cjs).

**Comment by IanVS** [comment]
This is finally ready for review.

File: code/lib/core-common/src/presets.ts
Comment (tmeasday) [reply code/lib/core-common/src/presets.ts:111]: Is this not a breaking change for any addons that have not made the changes we've made here in essentials?

> Additionally, this leaves it up to the consumer how to import/require the files, which means they'll get the right file (esm or cjs).

Can we keep this benefit without losing the absolute path (and keep stripping the absolute path in the vite builder?)

**Review by tmeasday — CHANGES_REQUESTED** [review]
Can we make this a non-breaking change? (Or do I have it wrong)?

File: code/lib/core-common/src/presets.ts
Comment (IanVS) [reply code/lib/core-common/src/presets.ts:111]: For anyone else watching, Tom and I chatted about this, and came up with a few ideas I'll try out of how we can do this in a non-breaking way, though other existing "super-addons" would still be broken in the vite builder, most likely, and we may need to log a warning in those cases, if possible.

Comment (IanVS) [reply code/lib/core-common/src/presets.ts:111]: OK, I've turned the `previewAnnotations` of sub-addons into a tuple, containing both the bare import name (for vite) and the absolute path (for webpack and back-compat).  I also added a warning when we can detect that sub-addons are not being re-exported from the top-level addon.  It doesn't have any instructions for how to address it, but my hope is that if it comes up, either users or authors will ask in GitHub/Discord about it.  And once we decide on a final path forward for these kinds of addons, we can add a note in the migration docs, and point to that.

**Comment by IanVS** [comment]
@Dschungelabenteuer and I found a few bugs in SSv6 that this now takes care of, and we've confirmed that his project (using pnpm and SSv7) is working correctly without any workarounds with these changes.  \o/

File: code/lib/builder-vite/src/codegen-iframe-script.ts
Comment (IanVS) [inline code/lib/builder-vite/src/codegen-iframe-script.ts:32]: SSv6 with the vite-builder and pnpm has been broken for a while.  I figured I might as well fix it in this PR, but I can move it to a different one if preferred.  I'd like to make a separate PR soon to create a `vite-core` that each of the frameworks re-export from, to avoid all the duplication in the frameworks above.  I'll do that soon.

**Review by tmeasday — CHANGES_REQUESTED** [review]
Some nitpicking, but one important detail to cover.

**Review by tmeasday — APPROVED** [review]

**Comment by Dschungelabenteuer** [comment]
> @Dschungelabenteuer and I found a few bugs in SSv6 that this now takes care of, and we've confirmed that his project (using pnpm and SSv7) is working correctly without any workarounds with these changes. \o/

I can confirm this works perfectly against my pnpm monorepo. This is looking pretty good, especially since my environment has some abstraction around Vite+Storybook configuration, which is shared across different Storybook instances and consumed as a local package. Basically, my setup could easily have raised some edge-case issues, but it did not at all! Well done @IanVS! This lets me get rid of ~7 "explicit" installs from my `package.json`/`.pnpmfile.cjs`, it feels like going out the hairdresser (I'm bald, though) :100: 

There are still some workarounds needed for MDX (which relies on react dependencies), but this is very likely another issue and IIRC this was mentioned during the monthly meeting. Not a big deal!

Please let me know if the latest changes require additional testing on my side!

**Comment by tmeasday** [comment]
Sorry @IanVS I was going to double check with @shilman before merging this yesterday but then it slipped our attention. Will do today!

**Review by shilman — APPROVED** [review]
This LGTM but is all a bit clever/subtle. I hope that we can come up with something simpler in 8.0!!

**Comment by IanVS** [comment]
I think the main change here is that if an addon wants to include other addons without forcing the user to also install those, it should re-export from them and point to its re-exports, rather than just specify the names of the sub-addons.  We'll need to decide if that's a good change to force community addons to make, or if there's another approach like all addons being imported into a preview.js or something.  This is "backwards compatible" for now in the sense that webpack will continue working as before (I think that's the "clever" bit you're referring to?), but the vite builder will be broken with pnp(m) for any community addons that have sub-addons.  So yeah, I hope we can refine this as we go.

**Comment by shilman** [comment]
@IanVS I am referring to the whole absolute path / bare thing. And the amount of work that an addon has to do to include other addons. I hope we can streamline this with the `preview.js` `addons` proposal.

That said, you've done an incredible job getting this working for 7.0 given the limitations of our current presets setup. Did not intend to piss on your work -- this is really great.

**Comment by IanVS** [comment]
>the amount of work that an addon has to do to include other addons

That should just really be a matter of specifying something like:

```
export const addons = [
	'super-addon/reexport',
	'super-addon/other-reexport'
];
```

I got a little fancy with export maps here, but there are other ways to accomplish that as well (but this worked the best with our tooling).

**Comment by IanVS** [comment]
@matsko could you open a new issue with all the details, and ideally a reproduction, and ping me on it?

**Pull request merged**
